### PR TITLE
Fixes #2022 - Illegal character 0x20 in state=HEADER

### LINKED ISSUE #2022: Illegal character 0x20 in state=HEADER for buffer
Hi,
we are using the the jetty HTTP client in our server to request external resources. But we get some strange errors on some pages since they do not behave compatible to the latest HTTP specs. 

For instance if you request [https://us.braun.com/en-us](https://us.braun.com/en-us) we get following logs:
`Illegal character 0x20 in state=HEADER for buffer DirectByteBuffer@7b362163[p=590,l=13192,c=16384,r=12602]={HTTP/1.1 200 OK\r\n...-Allow-Headers <<<: Origin, X-Reque...4E\xFd\xF0\xF84z \xF4+|\xF4\x95=\x1b>>>\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00...\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00}`

The Response headers looks as the following:
```
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store
Pragma: no-cache
Content-Type: text/html; charset=utf-8
Expires: -1
Vary: User-Agent, Accept-Encoding
Server: Microsoft-IIS/8.5
Set-Cookie: braun_us#lang=en-US; path=/
Set-Cookie: ASP.NET_SessionId=pimb44eu1ts4sd1raqs2sl4t; path=/; HttpOnly
Strict-Transport-Security: max-age=300
X-Frame-Options: SAMEORIGIN
Set-Cookie: __CSRFCOOKIE=m71fgPdokWdedn4Lt3TyntwmLA5QtPDEiFIXZRVByhalmgOTVboj5BP; path=/; HttpOnly
Access-Control-Allow-Headers : Origin, X-Requested-With, Content-Type, Accept
Access-Control-Allow-Origin: *
X-Frame-Options: SAMEORIGIN
X-UA-Compatible: IE=edge
X-Server: WPSC38
X-Content-Type-Options: nosniff
Date: Tue, 05 Dec 2017 20:51:25 GMT
Set-Cookie: BIGipServer~MAN3-PNG-RD1~us.braun.com-SC-ares.com=rd1o00000000000000000000ffff0a8bc13co8651; path=/
Content-Encoding: gzip
Transfer-Encoding: chunked
```

Affected Jetty Version:
9.3 (Tested with 9.3.14.v20161028)
9.4 (Tested with 9.4.8.v20171121)

Note that the issue does not happen always on that webpage.

The error seems to be in the [HttpParser](https://github.com/eclipse/jetty.project/blob/d4061fcfeb5a959e67547404a1fdfa5a203680db/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java#L1290) since a space is not expected in the HTTP header. However, there exist Servers which misbehave here and I expect that the Jetty Client should communicate with those Servers.

Do you see a way how Jetty can handle this properly? I see that there already exists many compliance checks in the Parser. Can this be added too?

### PR DESCRIPTION
**Pull request by fbuecklers**
Ignore invalid chars in http header names, when the compliance mode is not RFC7230

### CODE REVIEW

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java [hunk 1]
```diff
@@ -1287,7 +1287,13 @@ else if (_endOfContent == EndOfContent.UNKNOWN_CONTENT)
                         break;
                     }
 
-                    throw new IllegalCharacterException(_state,b,buffer);
+                    //Ignore all invalid characters
+                    if (complianceViolation(RFC7230,"https://tools.ietf.org/html/rfc7230#section-3.2 Invalid token in header name"))
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: I don't think this is too forgiving.  We really only want to allow optional white space... in fact probably only spaces.   I think to be safe we are going to need another state  OWS_AFTER_NAME

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java [hunk 2]
```diff
@@ -38,7 +39,7 @@
 
 public class HttpReceiverOverHTTP extends HttpReceiver implements HttpParser.ResponseHandler
 {
-    private final HttpParser parser = new HttpParser(this);
+    private final HttpParser parser = new HttpParser(this, -1, HttpCompliance.RFC2616);
```
Comment (gregw) [inline jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java]: why are you forcing the compliance mode here?  There is a lookup in the default constructor

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java (hunk 1 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: Also, strictly speaking the compliance mode should be RFC2616 as the OWS was not allowed there, so it would only be accepted in LEGACY mode.

Eventually we will replace this mechanism with some more fine grained.

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java [hunk 3]
```diff
@@ -286,7 +287,7 @@ public void testFoldedField7230() throws Exception
         Assert.assertThat(_bad, Matchers.containsString("Header Folding"));
         Assert.assertNull(_complianceViolation);
     }
-    
+
```
Comment (gregw) [inline jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: while this is a good formatting change... I really don't like seeing unrelated reformatting in PRs of substance.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectorHttpClientTransport.java [hunk 4]
```diff
@@ -160,6 +160,7 @@ protected EndPoint newEndPoint(SelectableChannel channel, ManagedSelector select
         {
             SocketChannelEndPoint endp = new SocketChannelEndPoint(channel, selector, key, getScheduler());
             endp.setIdleTimeout(client.getIdleTimeout());
+            //TODO: make compliance mode configurable
```
Comment (gregw) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectorHttpClientTransport.java]: @sbordet any thoughts on how to do this?

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java (hunk 2 above)
Comment (fbuecklers) [reply jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java]: Hm yes, but we for instead use jetty as a server and a client in one application. Changing the parser mode to none strict per default is not really an option for us. Since it changes the behavior both for the client and the server at once.
And as the comments of the compliance mode state out, that RFC2616 is compatible down to http 0.9 where whitespaces weren't explicitly disallowed. Where LEGACY also make header parsing case sensitive, which is also not an option for us. 
In general, I think a client should be more weak and acceptable against none compliant HTTP servers. For instance, chrome accepts the response from the provided page of the issue and parses the HTTP header correctly.

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java (hunk 3 above)
Comment (fbuecklers) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: Sorry, I thought I have removed all unnecessary code reformattings.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java (hunk 1 above)
Comment (fbuecklers) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: As I understood it correctly RFC2616 is compliant to HTTP 0.9 isn't it? At least the comments state it [HttpCompliance.java](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java)

Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: Yes it is, but HTTP/0.9 has no headers so accepting HTTP/0.9 does not give grace for OWS after a header name.    That is a LEGACY mode - ie something that Jetty used to support but no longer does if it is either in strict RFC2616 or RFC7230 modes

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java (hunk 2 above)
Comment (gregw) [reply jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java]: I think we want to leave the default compliance mode to RFC7230, which is the current spec and obsoletes RFC2616.   However there should be an easy way to programatically set the compliance mode on the client... @sbordet ?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java (hunk 1 above)
Comment (fbuecklers) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: But does the Legacy mode distinct between headers written in different cases? Then the Legacy mode is not what we are looking for...

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java (hunk 2 above)
Comment (fbuecklers) [reply jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java]: I can try to implement that the compliance mode is configurable on the jetty client itself and delegate the mode down to the parser creation if you like.

Comment (gregw) [reply jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java]: Let's wait for guidance from @sbordet about how to do that, but it would be good to include that in this PR.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java (hunk 1 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: And there is the Rub with the current big compliance mode switch!  It's all or nothing, when we really need to ability to turn on an off individual aspects of compliance.  We plan to do this for Jetty-10, but it is too invasive to do in a 9.4 dot release.

**Comment by gregw** [comment]
@fbuecklers sorry but the support here is on an open source basis so "big issues for one of your customers" is not a motivation to hurry up.  We do have commercial support clients that do get priority service and you can see the "sponsored" label on issues.    However we do support open source, so this is not a we are not going to do anything.   The fact that you are prepared to spend the effort making a PR means that we are happy to spend effort giving advise and getting it right so it can be merged.

So how to proceed? There are two things - how we accept the OWS and how we turn that mode on/off.

Firstly we need the state machine to be able to correctly ignore those OWS.  I'm pretty sure that is going to need another state called something like OWS_AFTER_NAME - that will handle space by staying in that state, colon by going to the value state and anything else is illegal state exception.    It would be good if you could have a go at implementing that and then adding unit tests that check for things like `Bad Header : value`.

For turning it on/off, we'll wait for comment from @sbordet.  Plus I'll have a think if we can incrementally change from a single mode enum, to an EnumSet of multiple compliance features.

**Comment by fbuecklers** [comment]
Okay, I can implement the new state, but what makes the space char so special compared to other invalid chars. I currently didn't get the reason why only a space will then be allowed, there may exist other servers which may send different invalid chars in their header names or is the space really a special case here? At least a have seen some invalid chars in Cookie strings for some reason...

**Comment by gregw** [comment]
I'm saying only spaces because we know what to do with spaces... treat them like optional white space.
For other illegal characters, we don't know what to do? are they part of the name? are they a divider to the value?  or are they some evidence of an attack or incorrect connection!

Flexibility is good, but too much flexibility can be a security risk

**Comment by fbuecklers** [comment]
But is this also true for a client? For a server, it is true that it will be a security attack. But for clients?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java (hunk 1 above)
Comment (fbuecklers) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: Okay I got it, then I will add the behavior on the legacy mode for now. Is there any documentation what will behave differently when we activate the Legacy mode per default? There are some more checks in the sources and I didn't understand them all for now...

**Comment by gregw** [comment]
Clients can be attacked by malicious servers just as servers can be attacked by malicious clients.

I'm all for flexibility, but in this case we are not going to ignore arbitrary characters other than what would be ignored if OWS was allowed there.

**Comment by fbuecklers** [comment]
Okay, if you find some time I'm happy if you can review my pull request and provide some feedback

**Comment by gregw** [comment]
@fbuecklers looking much better.  I'll take from here, so I've created #2064 so I can work on improving the compliance mode selection mechanism.

**Comment by joakime** [comment]
@gregw did you mean to close this?

**Comment by fbuecklers** [comment]
Cool thanks!
